### PR TITLE
errors: Add InvalidAppCommand errors for non-existent and not-found

### PR DESCRIPTION
**Pull request by clobrano**
Add a specific errors for

- command app not found
- command app not executable

LP: #1727425

- [x] Have you followed the [guidelines for contributing](https://github.com/snapcore/snapcraft/blob/master/CONTRIBUTING.md)?
- [x] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?
- [x] Have you successfully run `./runtests.sh static`?
- [x] Have you successfully run `./runtests.sh tests/unit`?

-----

### CODE REVIEW

**Comment by clobrano** [comment]
As far as I can see, the code that is traversed using the `LEGACY` adapter only raises `meta_errors.CommandError` (because  `_find_bin` raises only `subprocess.CalledProcessError`), for this reason I am able to differentiate the error between "file not found" and "file not executable" only for the `FULL` adapter.

**Comment by sergiusens** [comment]
Hi there, thanks for the PR. Not sure you noticed, but static tests have failed

**Comment by clobrano** [comment]
Sorry, I misread the results. I fixed the error in the files changed by this PR, what about the other files that fail (`lifecycle/_packer.py`, `multipass/test_instance_info.py` and `grammar_processing/test_global_grammar_processor.py`)?

File: snapcraft/internal/errors.py [hunk 1]
```diff
@@ -175,6 +175,32 @@ def __init__(self, command, app_name):
...
+        super().__init__(command=command, app_name=app_name)
+
+
+class InvalidAppCommandNotExecutable(SnapcraftError):
+
+    fmt = (
+        "Failed to generate snap metadata: "
+        "The specified command {command!r} defined in the app {app_name!r} does "
+        "not exist.\n"
+        "Ensure that {command!r} is relative to the prime directory."
```
Comment (sergiusens) [inline snapcraft/internal/errors.py]:
```suggestion
        "Ensure that {command!r} is executable."
```

**Comment by clobrano** [comment]
I thought about adding the new error classes to the test, I just didn't know how to refer to an existent, but not executable file in the test description

File: snapcraft/internal/errors.py [hunk 2]
```diff
@@ -175,6 +175,32 @@ def __init__(self, command, app_name):
         super().__init__(command=command, app_name=app_name)
 
 
+class InvalidAppCommandNotFound(SnapcraftError):
+
+    fmt = (
+        "Failed to generate snap metadata: "
+        "The specified command {command!r} defined in the app {app_name!r} does "
+        "not exist.\n"
+        "Ensure that {command!r} is relative to the prime directory."
```
Comment (cjp256) [inline snapcraft/internal/errors.py]: From a user perspective, I'm not sure what the 'prime' directory implies... especially as the building with lxd/multipass really hides the prime directory from the user (is there a better way than `snapcraft stage --shell; ls prime/`?)

File: snapcraft/internal/errors.py [hunk 3]
```diff
@@ -175,6 +175,32 @@ def __init__(self, command, app_name):
...
+
+    def __init__(self, command, app_name):
+        super().__init__(command=command, app_name=app_name)
+
+
+class InvalidAppCommandNotExecutable(SnapcraftError):
+
+    fmt = (
+        "Failed to generate snap metadata: "
+        "The specified command {command!r} defined in the app {app_name!r} does "
```
Comment (cjp256) [inline snapcraft/internal/errors.py]: What about dropping the "does not exist" portion and perhaps say "does exists, but is not executable"?

**Comment by cjp256** [comment]
The tests are failing on the ROS bits - I'm guessing that if this PR is rebased on master, that should fix it.

File: snapcraft/internal/errors.py (hunk 2 above)
Comment (sergiusens) [reply snapcraft/internal/errors.py]: `snapcraft prime --shell-after` is better until we have introspection,
`find prime` will also be more appropriate

File: snapcraft/internal/errors.py (hunk 3 above)
Comment (clobrano) [reply snapcraft/internal/errors.py]: Absolutely, I will change that

**Comment by sergiusens** [comment]
@cjp256 the ROS bits are a new error

File: snapcraft/internal/errors.py (hunk 2 above)
Comment (cjp256) [reply snapcraft/internal/errors.py]: Perhaps we can suggest that to the user?  Or state along the lines of "Ensure that {command!r} is installed with the correct path."

**Comment by sergiusens** [comment]
related bugs
https://bugs.launchpad.net/snapcraft/+bug/1797378
https://bugs.launchpad.net/snapcraft/+bug/1830520
https://bugs.launchpad.net/snapcraft/+bug/1797795
https://bugs.launchpad.net/snapcraft/+bug/1807117
https://bugs.launchpad.net/snapcraft/+bug/1808043
https://bugs.launchpad.net/snapcraft/+bug/1811362

File: snapcraft/internal/errors.py (hunk 2 above)
Comment (sergiusens) [reply snapcraft/internal/errors.py]: @clobrano hi, I just noticed the error message doesn't follow the pattern we have across the board, can you change the `: ` to `.\n`

we are planning a general change for error messages, but let's not go with to patterns at once.
Applying @cjp256's suggestion would be nice too.

Comment (clobrano) [reply snapcraft/internal/errors.py]: Ah, I didn't notice the pattern. Sure I'll do both changes

Comment (clobrano) [reply snapcraft/internal/errors.py]: Sorry for the delay, shall I convert all the occurences of `:` in `.\n`? Or maybe only in the error messages I changed? I see that this pattern is not observed in this file.

**Comment by sergiusens** [comment]
@clobrano hi there, any updates on this one?

**Comment by clobrano** [comment]
Hi @sergiusens, I was waiting for a reply about this question

https://github.com/snapcore/snapcraft/pull/2539#discussion_r297714675

File: snapcraft/internal/errors.py (hunk 2 above)
Comment (cjp256) [reply snapcraft/internal/errors.py]: @clobrano You are right, your pattern seems consistent. Just swap `"Ensure that {command!r} is relative to the prime directory."` with `"Ensure that {command!r} is installed with the correct path."`

File: snapcraft/internal/errors.py (hunk 3 above)
Comment (cjp256) [reply snapcraft/internal/errors.py]: Looks good.

**Comment by sergiusens** [comment]
Great, we can merge this once our release window re-opens

**Pull request merged**